nucliadb_vectors2/src/vectors/data_point/ops_hnsw.rs

```diff
@@ -251,24 +250,26 @@ impl<'a, DR: DataRetriever> HnswOps<'a, DR> {
                 let SearchValue {
                     neighbours: layer_res,
                     ..
-                } = self.layer_search(x, hnsw.get_layer(crnt_layer), 1, NO_FILTER, &entry_points);
+                } = self.layer_search(x, hnsw.get_layer(crnt_layer), 1, &entry_points);
                 neighbours = layer_res;
                 crnt_layer -= 1;
             }
             let entry_points: Vec<_> = neighbours.into_iter().map(|(node, _)| node).collect();
-            let result = self.layer_search(
-                x,
-                hnsw.get_layer(crnt_layer),
-                k_neighbours,
-                with_filter,
-                &entry_points,
-            );
+            let layer = hnsw.get_layer(crnt_layer);
+            let result = self.layer_search(x, layer, k_neighbours, &entry_points);
+            let mut solution = result.neighbours.iter().copied().map(|v| v.0).collect();
             let neighbours = result
                 .neighbours
                 .into_iter()
-                .map(|(n, _)| self.closest_up_node(hnsw.get_layer(0), n, x))
-                .filter(Option::is_some)
-                .flatten()
+                .flat_map(|(n, _)| {
+                    self.closest_up_node(
+                        &mut solution,
+                        with_filter.iter().copied(),
```

Comment:
This copy can be avoided in the `closest_up_node` method takes a slice as argument